### PR TITLE
Onboard SD hangs on "Loading..."

### PR DESCRIPTION
**Pull request by kisslorand**
### Requirements

BTT TFT

### Description

Many users reported they cannot access the onboard SD, TFT hangs with "Loading..." message or they got an "Error" popup.

### Benefits

Hopefully resolves the above.

### Related Issues

#1294, #1307, #1297

### CODE REVIEW

**Review by traffic-light — CHANGES_REQUESTED** [review]
Good job if this is working.

Can you take a look at my comments?

File: TFT/src/User/Menu/SendGcode.c
Comment (kisslorand) [reply TFT/src/User/Menu/SendGcode.c:3]: Because the commented part is replaced by the following lines. This is a test PR, I am waiting for feedback from others since I cannot test onboard SD. Once it is proven to resolve the intended issue I will clean it up and make another commit.

File: platformio.ini
Comment (kisslorand) [reply platformio.ini:31]: Yes, I can

**Comment by kisslorand** [comment]
PlatformIO.ini updated for temporary build fix.

Thanks @JulienThomas49 !!!

**Comment by oldman4U** [comment]
Great!

I will test it in an hour or so and let you know.

Thank you

File: TFT/src/User/Menu/SendGcode.c
Comment (traffic-light) [reply TFT/src/User/Menu/SendGcode.c:3]: Okay. Thank you for the explanation..

**Comment by traffic-light** [comment]
If you don't want to merge it, make it a draft

**Comment by kisslorand** [comment]
@traffic-light 
> If you don't want to merge it, make it a draft

Thanks for the hint. Marked a draft.

**Comment by oldman4U** [comment]
> PlatformIO.ini updated for temporary build fix.
> 
> Thanks @JulienThomas49 !!!

Would this temp fix allow you to compile the other two PR's also?

**Comment by oldman4U** [comment]
And is this somehow related to the compile issue?

https://github.com/MarlinFirmware/Marlin/commit/753cf994b6d529da9c8dd710a24a96b0388425ee

**Comment by digant73** [comment]
@kisslorand 
the ld spec files are reported under buildroot\ldscripts directory.
With the exception of stm32f4xx (based on  stm32f4xx_0x8000_iap.ld) all the other TFT variants have 48K of RAM. They primary differs for HEAP, STACK minumum reserved space. The 65K you see in vscode is not correct. The problem with the freeze in some TFT variants (such as BTT TFT35 v3 E3 etc...) seems releated to the extra resources they need (e.g. 4 serials instead of 3 etc...).
Try to find the code displayng the "Loading" message and verify if after that code a sort of malloc is executed. Normally the malloc used in the project have a wrong logic like "while(!malloc(....)). Maybe the freeze is caused by that loop (no looop exit because the malloc always fails (due to no more heap available)

**Comment by kisslorand** [comment]
According to the [datasheet](https://www.st.com/resource/en/datasheet/cd00220364.pdf) of STM32F107 its RAM size is 64.

**Comment by digant73** [comment]
on mygcoefs.c try to check the value provided "malloc" at row 41, 97 and 100.

in case of value NULL, display an error message. Let Radek or oldman (if they are available to make that test) test the result.
At least you should see the error message just before the freeze

**Comment by digant73** [comment]
> According to the [datasheet](https://www.st.com/resource/en/datasheet/cd00220364.pdf) of STM32F107 its RAM size is 64.

probably, but it's not the limit used in the env. You normally have other constrants such as less than 256K for the binary (e.g. 256 - 24)

**Comment by kisslorand** [comment]
I also got to the mygcodesfs.c but couldn't go further because not having onboard SD to test.

**Comment by digant73** [comment]
> I also got to the mygcodesfs.c but couldn't go further because not having onboard SD to test.

I know, for that reason I asked to let radek and oldman (having the TFT affected by the problem) test it. simply provide in your PR (just for testing) the check on the 3 "malloc" and display an error message if NULL is returned.

**Comment by oldman4U** [comment]
I am testing it right now.

**Comment by oldman4U** [comment]
I tested with 1 and 30 .gcode files on the SD card.

With the SD card already inserted and inserted after the printer has started. Removed the card and reinserted and also tested selecting another storage before selecting the on board sd card and vice versa.

I could not find an issue, but it would be good to have this confirmed by another user if possible.

Thank you

**Comment by kisslorand** [comment]
@oldman4U Is M503 working? Can you test?

**Comment by oldman4U** [comment]
M503 givers me:
M27
Not SD Printing
OK P15 B3

HOPE this helps

Please let me know if you need anything else

**Comment by radek8** [comment]
I'll get to the printer in 7 hours.
Then I test and give feedback

**Comment by kisslorand** [comment]
> M503 givers me:
> M27
> Not SD Printing
> OK P15 B3
> 
> HOPE this helps
> 
> Please let me know if you need anything else

What about M503 S0?

**Comment by oldman4U** [comment]
Rcv: ok P15 B3

**Comment by radek8** [comment]
> Rcv: ok P15 B3

How many pages?

**Comment by oldman4U** [comment]
3, and the is the last and only entry on page 3.

**Comment by kisslorand** [comment]
How many pages were with M503?

**Comment by oldman4U** [comment]
M503 only is 4 pages, same last entry

**Comment by kisslorand** [comment]
> > PlatformIO.ini updated for temporary build fix.
> > Thanks @JulienThomas49 !!!
> 
> Would this temp fix allow you to compile the other two PR's also?

Yes, if you apply the same to PlatformIO.ini.

**Comment by radek8** [comment]
I tested this PR on TFT35 E3 V3.0
It works perfectly.
SD card on board is loading without problems.
You can even open folders. Before editing, the display froze when trying to open the folder on board.
The M503 command listing is also complete.
The M503 S0 is also no problem.
You can release it to merge.
Thank you for a good job

**Comment by oldman4U** [comment]
This is great. Thank you all!!

Do we want to take this case to think about how we can avoid such situations in the future? Maybe a bit more testing before something is ready to be merged?

**Comment by kisslorand** [comment]
It's a double-edge sword. If you remember, PR  #1276  was not merged for at least a weak, no bug was reported. In fact not much bugs are reported in PRs. It was a mass hysteria only after merging into the master. 
Maybe a new branch would be better, something like "pre-release", "slave", "playgorund", etc and we could pull request our ideas/fixes there.

**Comment by oldman4U** [comment]
What we are working on is more or less this playground. With the last official release months ago. Maybe we should stop adding new features, fix some of the known bugs and ask BTT to make an official release?

**Comment by blueeagle69** [comment]
@kisslorand 
Just to add my confirmation with @oldman4U.

The onboard SD file contents now read fine again, thank you!

**Comment by kisslorand** [comment]
Sorry for the delay, I was busy checking the new STM32 library (v10.00) bug. I managed to build with v10.0.0 but not ready yet to make a PR of it.

This PR is ready for merge. Thanks for all for helping with ideas./infos and testing.

**Comment by oldman4U** [comment]
No reason to be sorry. You did a great job.

Thank you

**Comment by oldman4U** [comment]
Thank you

BigTreeTech <notifications@github.com> schrieb am Do. 3. Dez. 2020 um 02:13:

**Comment by oldman4U** [comment]
I am sorry, but it is not completely over. The test with the TFT24/35 E3 v3, and both failed in parts.This means that I can access the on board SD card in case it is the first storage I select for a print after turning on the printer. In case I select another storage and try to print from on board SD card after that, I get a: "Echo: No SD Card" followed by one peep and a double peep and then I am back on the Print screen where I can select the storage I would like to print from. Removing and inserting the SD card again does not help and I have to restart the printer to be able to access the on board SD card again.

So it is a little bit exactly the opposite of what we had before.

BUT, this only happens on my test SKR v1.4 board and NOT on my production printer which has also a TFT35 E3 v3 but the mainboard is a SKR E3 DIP.

Someone out there with a SKR v1.4 mainboard who could try to reproduce this behaviour. Maybe it is really related to my not really fine tuned SKR v1.4 firmware.

**Comment by radek8** [comment]
I have SKR V1.4
I will test in the evening

**Comment by radek8** [comment]
It works well for me. now it doesn't matter in which order I display the SD cards, it will access the folders without any problems.

**Comment by oldman4U** [comment]
Ok. Thank you. The SKR board I use inly for testing. Only the TFT connected and power.

Glad it works!!!

Thank you for checking

**Comment by Guilouz** [comment]
@kisslorand I have a issue with screen stay on "Loading..."

I can access to onboard SD, run a print but if I cancel it, screen stay on "Loading..." screen. Need to reset screen.

**Pull request merged**
